In [ ]:
!pip3 -q install pyspark weaviate-client > /dev/null

In [34]:
!pip show pyspark

Name: pyspark
Version: 3.4.0
Summary: Apache Spark Python API
Home-page: https://github.com/apache/spark/tree/master/python
Author: Spark Developers
Author-email: dev@spark.apache.org
License: http://www.apache.org/licenses/LICENSE-2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: py4j
Required-by: 


In [ ]:
!sudo apt-get install scala

In [ ]:
!wget https://repo1.maven.org/maven2/io/weaviate/spark-connector_2.13/1.2.4/spark-connector_2.13-1.2.4.jar

In [ ]:
!mv /content/spark-connector_2.13-1.2.4.jar /usr/local/lib/python3.10/dist-packages/pyspark/jars

In [36]:
from pyspark.sql import SparkSession
import os
#.config("spark.jars",
 #       "weaviate-spark-connector-assembly-v0.1.2.jar" )
spark = (
    SparkSession.builder
    .master("local[*]")
    .appName("weaviate")
    .getOrCreate()
)

spark.sparkContext.setLogLevel("WARN")

In [ ]:
!wget https://storage.googleapis.com/sphere-demo/sphere.100k.jsonl.tar.gz

In [ ]:
!gunzip /content/sphere.100k.jsonl.tar.gz

In [ ]:
!tar -xvf /content/sphere.100k.jsonl.tar

sphere.100k.jsonl


In [ ]:
df = spark.read.load("sphere.100k.jsonl", format="json")

In [ ]:
import weaviate
from weaviate.embedded import EmbeddedOptions

client = weaviate.Client(embedded_options=EmbeddedOptions())

embedded weaviate is already listing on port 6666


In [ ]:
client.schema.delete_all()

client.schema.create_class(
    {
        "class": "Sphere",
        "properties": [
            {
                "name": "raw",
                "dataType": ["string"]
            },
            {
                "name": "sha",
                "dataType": ["string"]
            },
            {
                "name": "title",
                "dataType": ["string"]
            },
            {
                "name": "url",
                "dataType": ["string"]
            },
        ],
    }
)

Embedded weaviate wasn't listening on port 6666, so starting embedded weaviate again
Started /root/.cache/weaviate-embedded: process ID 16160


In [ ]:
df.limit(1500).withColumnRenamed("id", "uuid").write.format("io.weaviate.spark.Weaviate") \
    .option("batchSize", 200) \
    .option("scheme", "http") \
    .option("host", "localhost:8080") \
    .option("id", "uuid") \
    .option("className", "Sphere") \
    .option("vector", "vector") \
    .mode("overwrite").save()

Here's what happens behind the scenes when the client uses the embedded options in the instantiation call:

The client downloads a Weaviate release from GitHub and caches it
It then spawns a Weaviate process with a data directory configured to a specific location, and listening to the specified port (by default 6666)

The server's STDOUT and STDERR are piped to the client

The client connects to this server process (e.g. to http://127.0.0.1:6666) and runs the client code
After running the code (when the application terminates), the client shuts down the Weaviate process

The data directory is preserved, so subsequent invocations have access to the data from all previous invocations, across all clients using the embedded option.

https://ai.facebook.com/blog/introducing-sphere-meta-ais-web-scale-corpus-for-better-knowledge-intensive-nlp/

https://huggingface.co/facebook/dpr-question_encoder-single-nq-base

https://weaviate.io/blog/sphere-dataset-in-weaviate